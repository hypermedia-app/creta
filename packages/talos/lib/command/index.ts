export { put } from './put'
export { putVocabs } from './put-vocabs'

export interface Command {
  endpoint: string
  updateEndpoint?: string
  token?: string
  user?: string
  password?: string
}
