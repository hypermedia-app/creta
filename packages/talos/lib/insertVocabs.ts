import { Quad } from 'rdf-js'
import StreamClient, { StreamClientOptions } from 'sparql-http-client'
import { prefixes, vocabularies } from '@zazuko/rdf-vocabularies'
import { INSERT } from '@tpluscode/sparql-builder'
import { sparql } from '@tpluscode/rdf-string'
import $rdf from 'rdf-ext'
import DatasetExt from 'rdf-ext/lib/Dataset'

function toTriple({ subject, predicate, object }: Quad) {
  return $rdf.triple(subject, predicate, object)
}

function insertVocab(client: StreamClient) {
  return ([prefix, vocab]: [string, DatasetExt]): Promise<void> => {
    const namespace = $rdf.namedNode(prefixes[prefix])

    const insert = INSERT.DATA`GRAPH <${namespace.value}> {
  ${vocab.map(toTriple).toString()}
}`
    const query = sparql`DROP SILENT GRAPH <${namespace.value}>;\n${insert}`.toString()

    return client.query.update(query)
  }
}

export async function insertVocabs(options: StreamClientOptions): Promise<void> {
  const client = new StreamClient(options)

  const datasets = await vocabularies({ only: ['hydra', 'acl', 'as'] })

  await Promise.all(Object.entries(datasets).map(insertVocab(client)))
}
