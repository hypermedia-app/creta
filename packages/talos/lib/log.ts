import Debugger from 'debug'

export const log = Debugger('talos')
log.enabled = true

export const debug = log.extend('debug')
