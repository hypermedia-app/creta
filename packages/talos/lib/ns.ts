import namespace from '@rdfjs/namespace'

type TalosTerms =
  'resources' |
  'action' |
  'default' |
  'overwrite' |
  'skip' |
  'merge'|
  'environmentRepresentation' |
  'replace'

export const talosNs = namespace<TalosTerms>('urn:talos:')
