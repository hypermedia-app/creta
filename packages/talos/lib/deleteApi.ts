import { NamedNode } from 'rdf-js'
import fetch from 'node-fetch'
import { log } from './log'

interface DeleteApi {
  token?: string
  apiUri: NamedNode
}

export async function deleteApi({ apiUri, token } : DeleteApi) {
  if (token) {
    const res = await fetch(apiUri.value, {
      method: 'DELETE',
      headers: {
        Authorization: `System ${token}`,
      },
    })

    if (res.ok) {
      log('Reset hydra:ApiDocumentation')
    } else {
      log('Failed to reset hydra:ApiDocumentation: %s', await res.text())
    }
  } else {
    log('No System token provided. API restart may be necessary for changes to be applied')
  }
}
