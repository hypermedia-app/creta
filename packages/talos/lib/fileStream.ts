import fs from 'fs'
import { Readable } from 'stream'
import path from 'path'
import * as mime from 'mime-types'
import replaceStream from 'replacestream'
import { parsers } from '@rdfjs/formats-common'
import isAbsoluteUrl from 'is-absolute-url'
import { log } from './log'

function replacer(basePath: string, resourceUrl: string, s: string, e = s) {
  return (_: unknown, match: string) => {
    let absolute: string
    if (isAbsoluteUrl(match)) {
      return `${s}${match}${e}`
    }

    if (match.startsWith('/') && basePath !== '/') {
      absolute = basePath + match
    } else {
      absolute = new URL(match, resourceUrl).toString()
      if (!match.endsWith('/')) {
        absolute = absolute.replace(/\/?$/, '')
      }
    }

    const absoluteUrl = new URL(absolute, resourceUrl)
    if (absoluteUrl.pathname === '/' && basePath !== '/') {
      absoluteUrl.pathname = basePath
      absolute = absoluteUrl.toString()
    }

    return `${s}${absolute}${e}`
  }
}

const angleBracketTransform = (basePath: string, resourceUrl: string) => replaceStream(/<([^>]+)>(?=([^"\\]*(\\.|"([^"\\]*\\.)*[^"\\]*"))*[^"]*$)/g, replacer(basePath, resourceUrl, '<', '>'))
const jsonTransform = (basePath: string, resourceUrl: string) => replaceStream(/"([./][^"]+)"/g, replacer(basePath, resourceUrl, '"'))

const filePatchTransforms = new Map([
  ['text/turtle', angleBracketTransform],
  ['application/n-triples', angleBracketTransform],
  ['application/n-quads', angleBracketTransform],
  ['application/trig', angleBracketTransform],
  ['application/ld+json', jsonTransform],
])

export function getPatchedStream(file: string, cwd: string, api: string, resourceUrl: string): Readable | null {
  const relative = path.relative(cwd, file)
  const basePath = new URL(api).pathname
  const mediaType = mime.lookup(file)
  if (!mediaType) {
    log('Could not determine media type for file %s', relative)
    return null
  }

  let fileStream = fs.createReadStream(file)
  if (filePatchTransforms.has(mediaType)) {
    fileStream = fileStream.pipe(filePatchTransforms.get(mediaType)!(basePath, resourceUrl))
  }
  const parserStream: Readable | null = parsers.import(mediaType, fileStream, {
    baseIRI: resourceUrl,
  }) as any

  if (!parserStream) {
    log('Unsupported media type %s from %s', mediaType, relative)
  }

  return parserStream
}
