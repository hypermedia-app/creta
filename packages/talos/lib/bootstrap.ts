import path from 'path'
import { DatasetCore, NamedNode } from 'rdf-js'
import walk from '@fcostarodrigo/walk'
import StreamClient, { StreamClientOptions } from 'sparql-http-client'
import $rdf from 'rdf-ext'
import clownface from 'clownface'
import { ResourcePerGraphStore } from '@hydrofoil/knossos/lib/store'
import { hydra } from '@tpluscode/rdf-ns-builders/strict'
import { getPatchedStream } from './fileStream'
import { log } from './log'

type Options = StreamClientOptions & {
  api: string
  apiUri: NamedNode
  cwd: string
}

export async function bootstrap({ api, apiUri, cwd, ...options }: Options): Promise<void> {
  const store = new ResourcePerGraphStore(new StreamClient(options))

  for await (const file of walk(cwd)) {
    const relative = path.relative(cwd, file)
    const resourcePath = path.relative(cwd, file)
      .replace(/\.[^.]+$/, '')
      .replace(/\/?index$/, '')

    const url = resourcePath === ''
      ? encodeURI(api)
      : encodeURI(`${api}/${resourcePath}`)

    const parserStream = getPatchedStream(file, cwd, api, url)
    if (!parserStream) {
      continue
    }

    const dataset: DatasetCore = $rdf.dataset()
    try {
      for await (const quad of parserStream) {
        dataset.add(quad)
      }
    } catch (e: any) {
      log('Failed to parse %s: %s', relative, e.message)
      continue
    }

    const pointer = clownface({ dataset })
      .namedNode(url)
      .addOut(hydra.apiDocumentation, apiUri)

    if (await store.exists(pointer.term)) {
      log(`Replacing resource ${url}`)
    } else {
      log(`Creating resource ${url}`)
    }

    await store.save(pointer)
  }
}
