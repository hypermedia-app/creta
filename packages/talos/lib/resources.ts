import path from 'path'
import fs from 'fs'
import { NamedNode } from 'rdf-js'
import walk from '@fcostarodrigo/walk'
import $rdf from 'rdf-ext'
import type DatasetExt from 'rdf-ext/lib/Dataset'
import TermSet from '@rdfjs/term-set'
import clownface from 'clownface'
import { deleteMatch } from 'rdf-dataset-ext'
import { log, debug } from './log'
import { getPatchedStream } from './fileStream'
import { optionsFromPrefixes } from './prefixHandler'
import { talosNs } from './ns'

interface ResourceOptions {
  existingResource: 'merge' | 'overwrite' | 'skip'
  environmentRepresentation: 'default' | 'replace'
}

export async function fromDirectories(directories: string[], api: string): Promise<DatasetExt> {
  const validDirs = directories.filter(isValidDir)
  return validDirs.reduce(toGraphs(api), Promise.resolve($rdf.dataset()))
}

function toGraphs(api: string) {
  return async function (previousPromise: Promise<DatasetExt>, dir: string): Promise<DatasetExt> {
    let previous = await previousPromise
    const dataset = $rdf.dataset()

    debug('Processing dir %s', dir)

    for await (const file of walk(dir)) {
      const relative = path.relative(dir, file)
      const resourcePath = path.relative(dir, file)
        .replace(/\.[^.]+$/, '')
        .replace(/\/?index$/, '')

      const url = resourcePath === ''
        ? encodeURI(api)
        : encodeURI(`${api}/${resourcePath}`)

      const parserStream = getPatchedStream(file, dir, api, url)
      if (!parserStream) {
        continue
      }

      debug('Parsing %s', relative)
      const parsedResourceOptions: Partial<ResourceOptions> = { }
      parserStream.on('prefix', optionsFromPrefixes(parsedResourceOptions))

      const resources = new TermSet<NamedNode>()
      const resourceOptions = clownface({ dataset: previous, graph: talosNs.resources })
      try {
        for await (const { subject, predicate, object, ...quad } of parserStream) {
          const graph: NamedNode = quad.graph.equals($rdf.defaultGraph()) ? $rdf.namedNode(url) : quad.graph

          if (!resources.has(graph)) {
            debug('Parsed resource %s', graph.value)
          }
          resources.add(graph)
          dataset.add($rdf.quad(subject, predicate, object, graph))
        }
      } catch (e: any) {
        log('Failed to parse %s: %s', relative, e.message)
      }

      resources.forEach(id => {
        const action = parsedResourceOptions.existingResource || 'default'
        const environmentRepresentation = parsedResourceOptions.environmentRepresentation || 'default'
        const options = resourceOptions.node(id)
        options
          .deleteOut(talosNs.action, talosNs.default)
          .addOut(talosNs.action, talosNs(action))
          .deleteOut(talosNs.environmentRepresentation, talosNs.default)
          .addOut(talosNs.environmentRepresentation, talosNs(environmentRepresentation))

        if (options.has(talosNs.environmentRepresentation, talosNs.replace).terms.length) {
          previous = deleteMatch(previous, undefined, undefined, undefined, id)
        }
      })
    }

    previous.addAll(dataset)
    return previous
  }
}

function isValidDir(dir: string) {
  if (!fs.existsSync(dir)) {
    log('Skipping path %s which does not exist', dir)
    return false
  }
  if (!fs.statSync(dir).isDirectory()) {
    log('Skipping path %s which is not a directory', dir)
    return false
  }

  return true
}
