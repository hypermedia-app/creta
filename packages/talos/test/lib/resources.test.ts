import { DatasetCore } from 'rdf-js'
import path from 'path'
import { expect } from 'chai'
import namespace from '@rdfjs/namespace'
import toCanonical from 'rdf-dataset-ext/toCanonical'
import { talosNs } from 'talos/lib/ns'
import { fromDirectories } from '../../lib/resources'
import { debug } from '../../lib/log'

debug.enabled = true

const ns = namespace('https://example.com')

describe('@hydrofoil/talos/lib/resources', () => {
  describe('fromDirectories', () => {
    let dataset: DatasetCore

    beforeEach(async () => {
      const dirs = [
        path.resolve(__dirname, '../resources'),
        path.resolve(__dirname, '../resources.foo'),
        path.resolve(__dirname, '../resources.bar'),
      ]
      dataset = await fromDirectories(dirs, ns().value)
    })

    it('merges resources from multiple graph documents', function () {
      const resource = dataset.match(null, null, null, ns())

      expect(toCanonical(resource)).to.matchSnapshot(this)
    })

    it('merges resources from dataset and graph documents', function () {
      const resource = dataset.match(null, null, null, ns('/trig/users/john-doe'))

      expect(toCanonical(resource)).to.matchSnapshot(this)
    })

    it('merges resources from multiple dataset documents', function () {
      const resource = dataset.match(null, null, null, ns('/trig/users/jane-doe'))

      expect(toCanonical(resource)).to.matchSnapshot(this)
    })

    it('marks a resource for "default" by default', () => {
      const [{ object: action }, ...more] = dataset.match(ns(), talosNs.action, null, talosNs.resources)

      expect(action.value).to.eq('default')
      expect(more).to.be.empty
    })

    it('marks a resource for "merge" when prefix is used', () => {
      const [{ object: action }, ...more] = dataset.match(ns('/project/creta/user.group/admins'), talosNs.action, null, talosNs.resources)

      expect(action.value).to.eq('merge')
      expect(more).to.be.empty
    })
  })
})
