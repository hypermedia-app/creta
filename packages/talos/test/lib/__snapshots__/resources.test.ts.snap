exports["@hydrofoil/talos/lib/resources fromDirectories merges resources from multiple graph documents"] = "<https://example.com/#part> <http://schema.org/maxValue> \"100\"^^<http://www.w3.org/2001/XMLSchema#integer> <https://example.com> .\n<https://example.com/#part> <http://schema.org/minValue> \"10\"^^<http://www.w3.org/2001/XMLSchema#integer> <https://example.com> .\n<https://example.com> <http://schema.org/hasPart> <https://example.com/#part> <https://example.com> .\n<https://example.com> <http://schema.org/name> \"Bar environment\" <https://example.com> .\n<https://example.com> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Thing> <https://example.com> .\n";

exports["@hydrofoil/talos/lib/resources fromDirectories merges resources from dataset and graph documents"] = "<https://example.com/trig/users/john-doe> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Person> <https://example.com/trig/users/john-doe> .\n<https://example.com/trig/users/john-doe> <http://www.w3.org/ns/earl#test> \"trig\" <https://example.com/trig/users/john-doe> .\n<https://example.com/trig/users/john-doe> <http://xmlns.com/foaf/0.1/name> \"John Doe\" <https://example.com/trig/users/john-doe> .\n";

exports["@hydrofoil/talos/lib/resources fromDirectories merges resources from multiple dataset documents"] = "<https://example.com/trig/users/jane-doe> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Person> <https://example.com/trig/users/jane-doe> .\n<https://example.com/trig/users/jane-doe> <http://www.w3.org/ns/earl#test> \"trig\" <https://example.com/trig/users/jane-doe> .\n<https://example.com/trig/users/jane-doe> <http://xmlns.com/foaf/0.1/name> \"Jane Doe\" <https://example.com/trig/users/jane-doe> .\n";

exports["@hydrofoil/talos/lib/resources fromDirectories uses the last representation when is marked to replace other envs"] = "<https://example.com/only/one> <http://schema.org/name> \"bar\" <https://example.com/only/one> .\n";

exports["@hydrofoil/talos/lib/resources fromDirectories uses the last representation when is marked in another env"] = "<https://example.com/only/two> <http://schema.org/name> \"bar\" <https://example.com/only/two> .\n";

exports["@hydrofoil/talos/lib/resources fromDirectories uses the last representation when is marked in a dataset document"] = "<https://example.com/only/three> <http://schema.org/name> \"bar\" <https://example.com/only/three> .\n";

